Update AIP-193 with actionable errors guidance.

jschneider11 opened the pull request.

jschneider11 committed:
Update 0193.md

Incorporating writing guidance for messages into AIP-193.

jschneider11 committed:
Updated tables to Markdown.

jschneider11 committed:
Correcting Markdown on the first table only.

jschneider11 committed:
Removing pluses in first table for testing.

jschneider11 committed:
Updated JSON. Changed first table to bulleted list.

jschneider11 committed:
Fixed JSON, removed text referencing rtables. changed all tables to bulleted lists.

jschneider11 committed:
Committing changes again: Fixed JSON, text referring to tables, moved tables to bulleted lists.

jschneider11 committed:
Committing pre-formatted content from Andrei

jschneider11 committed:
Fixed indentation problem starting at line 207.

jschneider11 committed:
Merge branch 'master' into patch-1

jschneider11 committed:
Update 0193.md

Adding writing guidance and converting HTML to Markdown.

shwoodard committed:
Fix whitespace, especially line width.

* Restrict line width to 80 characters, applying standards for md
  format, e.g wrapping for links, etc.
* Add whitespace after head matter.
* Format bullets using dashes instead of astrices, to match style of the
  other AIPs.

andrei-scripniciuc committed:
Merge branch 'master' into gq-jschneider-patch-1

jschneider11 committed:
Merge pull request #1 from shwoodard/gq-jschneider-patch-1

Fix whitespace, especially line width.

shwoodard commented on aip/general/0193.md:
```diff
@@ -9,34 +9,51 @@ placement:
-your particular API. Additionally, error messages **must not** assume that the
-user will know anything about its underlying implementation.
+Services **must** return a
+[`google.rpc.Status`](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto)
```
`md` style: Use named anchors for consistency with other APIs, e.g.,

```
[`google.rpc.Status`][Status]
```

Instead of,

```
[`google.rpc.Status`](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto)
```

here and throughout.

shwoodard replied:
This is fixed.

shwoodard committed:
Reestablish and add named anchors.

shwoodard committed:
Fix broken links and expand warning

jschneider11 committed:
Update 0193.md

Removed anchor for error-reason and accepting Sam's updates to bullets.

shwoodard committed:
Remove unused named anchor

jschneider11 committed:
Update 0193.md

Fixed bullets from 3-level to 2-level.

shwoodard commented on aip/general/0193.md:
```diff
@@ -75,132 +86,191 @@ following:
-language so that messages are easily searchable in common knowledge
-bases, such as StackOverflow&trade;.
+- `code int32`
+  - The status code, which **should** be an enum value of
```
Why is this a **should** and not a **must**?

jschneider11 replied:
The language needs to be reviewed by Stanis.

sgiovanni1 replied:
Agree with @shwoodard that this should be **must**.

jschneider11 replied:
Updated. I left the Localization text as **should** be in English. Please advise.

andrei-scripniciuc replied:
SGTM

shwoodard commented on aip/general/0193.md:
```diff
@@ -75,132 +86,191 @@ following:
+    resources available to fulfill the request. Try a different zone, or
+    try again later.\"
+- `details Object (repeated)`
+  - Text
```
I think "Text" here is confusing. What does it mean in this context? `details` is a repeated Object in JSON and a repeated Message in proto. I don't see what value the word, "Text", adds here. It is an overloaded term. e.g., some databases have a text column type.

jschneider11 replied:
Can we just remove the line with - Text?

sgiovanni1 replied:
I think it's still beneficial to have a description here to be consistent with the other bullet points. How about something along the lines of "Additional details about the error, which include ErrorInfo, LocalizedMessage, etc."?

jschneider11 replied:
Updated.

shwoodard commented on aip/general/0193.md:
```diff
@@ -75,132 +86,191 @@ following:
-machine-readable identifier _in addition to_ the `code` field, changing the
-error message **must** be considered a backwards-incompatible change.
+Services must return a
+[`google.rpc.Status`][Status] message when an API error occurs. For each
```
It looks like these two lines could be reformatted. The first line can take more before max width. This will need to cascade through the paragraph. Your editor should be able to assist.

jschneider11 replied:
I reformatted the entire file to meet max width.

shwoodard commented on aip/general/0193.md:
```diff
@@ -75,132 +86,191 @@ following:
-differentiate one error for another and even parse the error message to
-extract variables from dynamic segments.
+Structured details with machine-readable identifiers **must** be used if
+users will need to write code against a specific aspect of the error.
```
This makes it sound like variables in message segments only need to be included in ErrorInfo.metadata if the API author thinks that their API will be used to write code against. How will they know? This language **must** be much firmer. Any variables in messages **must** be in  the metadata.

sgiovanni1 replied:
For what it's worth, I believe this was existing content of the AIP. That said, I agree that the language is confusing.

@jschneider11 What about changing this to "Structured details with machine-readable identifiers must be used so that users can write code against specific aspects of the error."?

jschneider11 replied:
SGTM. Thanks.

jschneider11 replied:
Updated.

shwoodard commented on aip/general/0193.md:
```diff
@@ -75,132 +86,191 @@ following:
-message by adding [`google.rpc.LocalizedMessage`][LocalizedMessage] to
-[`Status.details`][Status details].
+For the purpose of this discussion, the ErrorDetails section contains
+three payloads: `ErrorInfo`, `LocalizedMessage`, and `HelpLink`.
```
This is not useful or true. It can contain anything and often does have BadRequest or PreconditionFailure, etc.

andrei-scripniciuc replied:
@shwoodard @sgiovanni1 would it make sense to change this to say:

> ErrorDetails **must** contain one instance of `ErrorInfo`, one `LocalizedMessage`, and one `HelpLink`, and **may** contain one instance of additional error detail types as defined by [google/rpc/error_details.proto.](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto).

Please correct me if I'm wrong on whether `LocalizedMessage` and `HelpLink` are required, I'm pretty sure `ErrorInfo` is.

sgiovanni1 replied:
@andrei-scripniciuc It seems to me that this sentence is merely an introduction to the following sections that describes `ErrorInfo`, `LocalizedMessage`, and `HelpLink`. Do we need to spell out what must be included? A different way to position this is something like below.

In the following section, we will describe `ErrorInfo`, `LocalizedMessage`, and `HelpLink` in more detail.

shwoodard replied:
+1

jschneider11 replied:
Updated. "The following section describes `ErrorInfo`, `LocalizedMessage`, and `HelpLink`
in more detail."

shwoodard commented on aip/general/0193.md:
```diff
@@ -75,132 +86,191 @@ following:
+    are unique within a particular domain of errors. The error reason
+    **must** do the following:
+  - Be at most 63 characters and match a regular expression of
+    \`\[A-Z\]\[A-Z0-9\_\]+\[A-Z0-9\]\`, which represents
```
This regex is not written correctly. It does not render property with the site generator. Please fix.

jschneider11 replied:
@andrei-scripniciuc 
Should we use the same regex that you (Andrei) provided for the metadata string?

shwoodard replied:
I already fixed this.

shwoodard commented on aip/general/0193.md:
```diff
@@ -75,132 +86,191 @@ following:
+  - Be meaningful enough for a human reader to understand what the
+    reason refers to.
+  - Map to one and only one cause, That is, avoid making machine actors
+    reliant on ErrorInfo.
```
I don't know what this means. Please clarify.

andrei-scripniciuc replied:
I think this is trying to say don't put dynamic data in here that can't be read from `error_info.metadata`.

@shwoodard should this say that any dynamic values must be present in `error_info.metadata`? It feels a bit odd to define it that way but [LocalizedMessage](https://github.com/googleapis/googleapis/blob/13c6878a4c2822c6db7b986e717bf57943bf1681/google/rpc/error_details.proto#L277) doesn't have its own metadata value.

I think the bigger question here is, what's the purpose of `LocalizedMessage` relative to `ErrorInfo`? Specifically, what should `error_info.message` be used for, and what should `localized_message.message` be used for? The latter I think is what we want to display to users, but is there any use for the former, except for legacy applications?

jschneider11 replied:
From the text "The LocalizedMessage payload should contain the complete resolution to the error."
Shouldn't this be a "must?"

andrei-scripniciuc replied:
Please remove this statement as it's confusing and does not add anything, this above statement is actually sufficient:

>   + Be unique and consumable by machine actors for automation.

jschneider11 replied:
Done.

andrei-scripniciuc replied:
Sorry for the confusion, what I meant was that this statement is the one that's sufficient and we should keep:

> Be unique and consumable by machine actors for automation.

This statement, that's highlighted by this comment, is the one we should remove:

>  - Map to one and only one cause, That is, avoid making machine actors
    reliant on ErrorInfo.

shwoodard commented on aip/general/0193.md:
```diff
@@ -75,132 +86,191 @@ following:
+    in length. When identifying the current value of an exceeded limit,
+    the units **should** be contained in the key, not the value.
+  - *Example*: `"zone": "us-east1-a", "vmType": "e2-medium", "attachment":
+     "local-ssd=3,nvidia-t4=2"`
```
the rendering of these key value pairs is not indicative of a map with the site-generator. Please fix.

andrei-scripniciuc commented on aip/general/0193.md:
```diff
@@ -210,75 +302,291 @@ map for the error payload to be backwards compatible, even if the value
+- Place the Resolution in a newline after a line space
+- Include a hyperlink for documentation as part of the description
+
+The following example shows an actionable error message for the gcloud
```
this content should be moved to Cloud SDK AIPs (https://google.aip.dev/cloud#sdk), but I can do that in a follow-up CL to get this unblocked.

andrei-scripniciuc replied:
for now I suggest removing this paragraph from here to keep this AIP generic.

jschneider11 replied:
I removed the entire Message Layout section, as it wouldn't make sense if we only removed parts of it.

shwoodard commented on aip/general/0193.md:
```diff
@@ -210,41 +280,237 @@ map for the error payload to be backwards compatible, even if the value
+    bcp47](https://www.rfc-editor.org/rfc/bcp/bcp47.txt) (Tags for
+    Identifying Languages).
+  - *Example*: "en-US" , "fr-CH", "es-MX"
+- `message string`
```
I would call out that dynamic segments--variables--in the Help.message also must be in ErrorInfo.metadata. Otherwise, users will parse it and it will break clients to change the format of the string.

jschneider11 replied:
@andrei-scripniciuc 
Andrei, Agreed?
Do you mean Helplink.message?

andrei-scripniciuc replied:
Agreed, but the type is called `Help`, not `HelpLink`. Please update references to it in the doc. The type is defined here and you're already linking to it: https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto#L261

jschneider11 replied:
Done. Also, Sam's comment seems to be resolved by this sentence: "The message
    **should** include, where needed, data provided in other fields such
    as metadata."

shwoodard commented on aip/general/0193.md:
```diff
@@ -210,41 +280,237 @@ map for the error payload to be backwards compatible, even if the value
+##### Message alignment
+
+By default, `LocalizedMessage` is populated with the same message as
+`Status.message`. Should you choose, you can also present a different
```
What does this. Does the server have this behavior?

andrei-scripniciuc replied:
@jschneider11 this is an implementation detail, and should not be part of the AIP.

Perhaps you can re-word this to state that the Message is an english only error, and  LocalizedMessage:  is the translation, while stating that if the message needs to change, change LocalizedMessage instead.

jschneider11 replied:
Is there any case where the end-user message would need to differ from the debug message? 
The Localized message will also be in English until translation to other languages is provided.
Otherwise, do we remove this entire section? I'd run this by Stanis before doing so, just to make sure we're not removing guidance that he wanted here.

andrei-scripniciuc replied:
Sam's point is that we should not say "by default" unless that default behavior is actually enforced by something.

jschneider11 replied:
Removed "By default."

shwoodard commented on aip/general/0193.md:
```diff
@@ -210,41 +280,237 @@ map for the error payload to be backwards compatible, even if the value
+displaying messages to end users.
+
+The error message presented in `LocalizedMessage.message` **should** be
+the same as `Status.message` or it **should** be an alternate message.
```
This is a bit confusing it should be one or it should be another. This doesn 't seem logical.

sgiovanni1 replied:
+1. This was confusing to me too. The statement seems redundant. I'd vote for removal.

jschneider11 replied:
Done.

shwoodard commented on aip/general/0193.md:
```diff
@@ -210,41 +280,237 @@ map for the error payload to be backwards compatible, even if the value
+- The service plans to localize either immediately or in the near future. See,
+   "[Localization](#localization)".
+- This practice enables clients to find an error message consistently in
+  one location, `LocalizedMessaage.message`, across all methods of the
```
Again, what enforces this? What if there is only a Status.message and now LocalizedMessage.message.

andrei-scripniciuc replied:
If this is not enforced we should remove it.

shwoodard replied:
I agree.

jschneider11 replied:
Removed.

shwoodard commented on aip/general/0193.md:
```diff
@@ -210,41 +280,237 @@ map for the error payload to be backwards compatible, even if the value
+surfaces such as gcloud, Pantheon, or Terraform.
+
+- Place the Error description in the first line (required)
+- List each instance of metadata in a new line, indented
```
JSON is rendered by the server without whitespace. Please remove.

jschneider11 replied:
This appears to have been removed as part of the removal of the Layout section.

shwoodard commented on aip/general/0193.md:
```diff
@@ -210,41 +280,237 @@ map for the error payload to be backwards compatible, even if the value
+
+- Place the Error description in the first line (required)
+- List each instance of metadata in a new line, indented
+- Place the Resolution in a newline after a line space
```
What does this mean?

jschneider11 replied:
@andrei-scripniciuc Andrei, I removed the entire Message Layout section. It wouldn't make sense otherwise.

shwoodard commented on aip/general/0193.md:
````diff
@@ -210,41 +280,237 @@ map for the error payload to be backwards compatible, even if the value
+hardware configuration or at a later time. For more information, see the
+troubleshooting documentation:
+https://cloud.google.com/compute/docs/resource-error
+```
````
gCloud output is orthogonal to API design and should not be included here.

andrei-scripniciuc replied:
Agreed. @jschneider11 we will need to add a new AIP under https://google.aip.dev/cloud#sdk for this content, for now please remove it from this AIP as it's not cloud-specific.

jschneider11 replied:
Removed.

shwoodard commented on aip/general/0193.md:
```diff
@@ -253,32 +519,44 @@ it is difficult to disambiguate one error from another across an entire
 message, which would be a backwards incompatible change.
 
+
+### Updating Status.Message
```
Downcase "Message". it is a field. Uppercase is confusing.

jschneider11 replied:
Done.

shwoodard commented on aip/general/0193.md:
```diff
@@ -253,32 +519,44 @@ it is difficult to disambiguate one error from another across an entire
+
+### Updating Status.Message
+
+**Because Status.Message is part of the API contract**, avoid updating
```
Same here. "Message" in "Status.Message" should be "message".

jschneider11 replied:
Done.

shwoodard commented on aip/general/0193.md:
```diff
@@ -253,32 +519,44 @@ it is difficult to disambiguate one error from another across an entire
+it entirely in favor of updating LocalizedMessage; for example, clients
+often perform string matches on the text to differentiate one error for
+another and even parse the error message to extract variables from
+dynamic segments.
```
This doesn't take into account that this behavior, on the part of clients, should not occur if proper ErrorInfo and metadata semantics are used.

andrei-scripniciuc replied:
Agreed, this should call out that clients need to perform string matches if the only option they have for extracting dynamic information is the message itself. If structured data is present in metadata, they can key into that instead and avoid this.

jschneider11 replied:
How about adding the following to the existing text?

Clients **must** only perform string matches if their only option for extracting dynamic information is the message itself.  If structured data is present in metadata, the recommended practice is to key into that data instead.

shwoodard replied:
LGTM

jschneider11 replied:
Updated in next commit.

shwoodard reviewed (commented):
There are a lot of **should**s and **should not**s in here. Each should be evaluated for if they *can* be a **must** or **must not**. When possible, it is better to be restrictive.

jschneider11 committed:
Updated bullets to Markdown

Updated bullets to Markdown

jschneider11 committed:
Update 0193.md

Made the following updates:

- Fixed Text field for details.Object (repeated)
- Updated "Structure details with machine-readable identifiers."
- Removed the entire "Message Layout" section, as Andrei will update the SDK AIP instead.
- All Status.message fields were already downcased.

-

andrei-scripniciuc commented on aip/general/0193.md:
```diff
@@ -75,224 +84,354 @@ following:
-message in the `details` field, but there **may** be a `BadRequest` and
-a [`PreconditionFailure`][PreconditionFailure].
++ This `message` string is considered a problem description. It is intended for
+  human developers to understand the problem and is more detailed than
```
I would just say "developers" instead of "human developers"

jschneider11 replied:
Ack. Updated.

andrei-scripniciuc commented on aip/general/0193.md:
```diff
@@ -75,224 +84,354 @@ following:
+  + A developer-facing error message, which **should** be in English.
+    Any user-facing error message **should** be located in the
+    LocalizedMessage payload.
+  + *Example message*: \"The zone \'us-east1-a\' does not have enough
```
backticks are unnecessary here.

jschneider11 replied:
Updated.

andrei-scripniciuc replied:
can you also update the sample error on line 53 with the same change?

andrei-scripniciuc commented on aip/general/0193.md:
```diff
@@ -75,224 +84,354 @@ following:
+  + A short snake_case description of why the error occurred. Error reasons are
+    unique within a particular domain of errors. The error reason **must** do the following:
+  + Be at most 63 characters and match a regular expression of
+    \`\[A-Z\]\[A-Z0-9\_\]+\[A-Z0-9\]\`, which represents UPPER_SNAKE_CASE.
```
This regex should just be `[A-Z][A-Z0-9_]+[A-Z0-9]`. Different languages have different escape characters.

jschneider11 replied:
Updated.

andrei-scripniciuc commented on aip/general/0193.md:
```diff
@@ -75,224 +84,354 @@ following:
+  + Be unique and consumable by machine actors for automation.
+  + Be meaningful enough for a human reader to understand what the reason 
+    refers  to.
+  + Map to one and only one cause, That is, avoid making machine actors reliant
```
should That is be capitalized here if it's after a comma?

jschneider11 replied:
Good catch. Thanks. Updated.

andrei-scripniciuc commented on aip/general/0193.md:
```diff
@@ -75,224 +84,354 @@ following:
++ `metadata map`
+  + Additional structured details about this error, which **should** provide
+    important context for customers to identify resolution steps.  Keys
+    **should** match /\[a-zA-Z0-9-\_\]/ and be limited to 64 characters in
```
the regex here seems wrong, "us-", "us_", "us-_" are all valid keys according to it. I can help craft a better one, what are the constraints here?

jschneider11 replied:
@sgiovanni1  It looks like the constraints for metadata are using upper or lower case letters, or numbers, Stanis, can you please verify this?

jschneider11 replied:
@andrei-scripniciuc Should we add the wording that metadata keys must start with a lowercase letter and be at least two characters?

andrei-scripniciuc replied:
the regex here must match the one in the proto. please use `/[a-z][a-zA-Z0-9-_]+/`, we're in the process of updating the public proto documentation to match this: https://github.com/googleapis/googleapis/blob/fe20507f2a5a17da501cd3bcd4dfbc7675fb5a3f/google/rpc/error_details.proto#L69

jschneider11 replied:
Updated.

shwoodard committed:
Whitespace and minor edits

* Fix all whitespace--vertical and horizontal
  * Line length
  * Remove unneeeded vertical spacing between headers and paragraphs
* Edits
  * Change error messages from code treatment to quoted text
  * Minor rephrasing here and there

shwoodard committed:
Fix all regex and more whitespace fixes.

jschneider11 committed:
Update 0193.md

 - Changed "human developers" to "developers."
- Removed capitalization for "That."
- Added spaces between words that appear next to each other.
- Updated guidance on how to extract dynamic information.

shwoodard committed:
Fix whitespace

jschneider11 committed:
Update 0193.md

- Simplified last line of introduction in ErrorDetails.
- Fixed regex in metadata map section.

Note: If there's any way to outdent domain string and metadata map in the bulleted list, it would make the text read better.

shwoodard committed:
Another whitespace fix

jschneider11 committed:
Resolved hidden comments in the PR

- Removed confusing verbiage about LocalizedMessage being the same as Status.message or different.
- Removed section stating reasons to present the same error message in both locations.

jschneider11 committed:
Moved reasons to present differing or non-differing messages to the Rationale section

Moved reasons to present the same error message in Status.message and Localized.message vs. differing messages to the Rationale section.

jschneider11 committed:
Fixing line formatting.

Fixing line formatting on lines 392 and 475.

jschneider11 committed:
Closed 3 comments, changed state to Draft

 - Changed the state of the AIP to Draft as suggested by Andrei.
- Updated all references to HelpLink to Help.
- Removed line 192, which duplicated line 195+
- Removed "By default" line 389 as it is not enforced by anything.
-

jschneider11 committed:
Reversed line about uniqueness.  Fixed typos.

 - Fixed formatting issue on line 337
- Fixed typo on line 347
- Added back in: "Be unique and consumable by machine actors for automation." Removed: Map to one and only one cause."

andrei-scripniciuc commented on aip/general/0193.md:
```diff
@@ -75,132 +84,190 @@ following:
++ `code int32`
+  + The status code, which **must** be an enum value of
+    [`google.rpc.Code`][Code].
+  + *Example*: 429
```
this example is wrong, google.rpc.Code is different from HTTP status code. In this case, it should be `RESOURCE_EXHAUSTED`: https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto#L108

jschneider11 committed:
Replace example error code for status example

The example message is actually RESOURCE_EXHAUSTED, code 8, HTTP status code information.

jschneider11 committed:
Fixed error code in JSON example

Fixed error code in JSON example to reflect the RESOURCE_EXCEEDED code of 8.

andrei-scripniciuc committed:
Merge branch 'master' into patch-1
